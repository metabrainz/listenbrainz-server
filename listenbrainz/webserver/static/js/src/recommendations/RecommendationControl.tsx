import * as React from "react";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { IconDefinition } from "@fortawesome/fontawesome-common-types"; // eslint-disable-line import/no-unresolved
import { IconProp } from "@fortawesome/fontawesome-svg-core";

export type RecommendationControlProps = {
  cssClass: string;
  action: (event: React.SyntheticEvent) => void;
  iconHover: IconDefinition;
  icon: IconDefinition;
  title: string;
};

const RecommendationControl = (props: RecommendationControlProps) => {
  const { iconHover, icon, action, title, cssClass } = props;
  return (
    <div
      className={`recommendation-icon ${cssClass}`}
      title={title}
      onClick={action}
      onKeyPress={action}
      role="button"
      tabIndex={0}
    >
      <span className="on">
        <FontAwesomeIcon icon={iconHover as IconProp} />
      </span>
      <span className="off">
        <FontAwesomeIcon icon={icon as IconProp} />
      </span>
    </div>
  );
};

export default RecommendationControl;
