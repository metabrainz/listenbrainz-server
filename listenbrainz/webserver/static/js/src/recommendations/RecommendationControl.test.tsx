import * as React from "react";
import { mount } from "enzyme";

import { faMeh } from "@fortawesome/free-solid-svg-icons";

import { faMeh as faMehRegular } from "@fortawesome/free-regular-svg-icons";
import RecommendationControl, {
  RecommendationControlProps,
} from "./RecommendationControl";

// Font Awesome generates a random hash ID for each icon everytime.
// Mocking Math.random() fixes this
// https://github.com/FortAwesome/react-fontawesome/issues/194#issuecomment-627235075
jest.spyOn(global.Math, "random").mockImplementation(() => 0);

const props: RecommendationControlProps = {
  cssClass: "bad_recommendation",
  action: () => {},
  iconHover: faMeh,
  icon: faMehRegular,
  title: "This is a bad recommendation",
};

describe("RecommendationControl", () => {
  it("renders correctly", () => {
    const wrapper = mount(<RecommendationControl {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.props().cssClass).toEqual("bad_recommendation");
    expect(wrapper.props().title).toEqual("This is a bad recommendation");
    expect(wrapper.props().icon).toEqual(faMehRegular);
    expect(wrapper.props().iconHover).toEqual(faMeh);
  });
});
