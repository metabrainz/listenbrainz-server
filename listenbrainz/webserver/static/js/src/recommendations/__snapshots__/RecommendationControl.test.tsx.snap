// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendationControl renders correctly 1`] = `
<RecommendationControl
  action={[Function]}
  cssClass="bad_recommendation"
  icon={
    Object {
      "icon": Array [
        496,
        512,
        Array [],
        "f11a",
        "M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 448c-110.3 0-200-89.7-200-200S137.7 56 248 56s200 89.7 200 200-89.7 200-200 200zm-80-216c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160-64c-17.7 0-32 14.3-32 32s14.3 32 32 32 32-14.3 32-32-14.3-32-32-32zm8 144H160c-13.2 0-24 10.8-24 24s10.8 24 24 24h176c13.2 0 24-10.8 24-24s-10.8-24-24-24z",
      ],
      "iconName": "meh",
      "prefix": "far",
    }
  }
  iconHover={
    Object {
      "icon": Array [
        496,
        512,
        Array [],
        "f11a",
        "M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm-80 168c17.7 0 32 14.3 32 32s-14.3 32-32 32-32-14.3-32-32 14.3-32 32-32zm176 192H152c-21.2 0-21.2-32 0-32h192c21.2 0 21.2 32 0 32zm-16-128c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z",
      ],
      "iconName": "meh",
      "prefix": "fas",
    }
  }
  title="This is a bad recommendation"
>
  <div
    className="recommendation-icon bad_recommendation"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
    title="This is a bad recommendation"
  >
    <span
      className="on"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              496,
              512,
              Array [],
              "f11a",
              "M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm-80 168c17.7 0 32 14.3 32 32s-14.3 32-32 32-32-14.3-32-32 14.3-32 32-32zm176 192H152c-21.2 0-21.2-32 0-32h192c21.2 0 21.2 32 0 32zm-16-128c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z",
            ],
            "iconName": "meh",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-meh fa-w-16 "
          data-icon="meh"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 496 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm-80 168c17.7 0 32 14.3 32 32s-14.3 32-32 32-32-14.3-32-32 14.3-32 32-32zm176 192H152c-21.2 0-21.2-32 0-32h192c21.2 0 21.2 32 0 32zm-16-128c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </FontAwesomeIcon>
    </span>
    <span
      className="off"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              496,
              512,
              Array [],
              "f11a",
              "M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 448c-110.3 0-200-89.7-200-200S137.7 56 248 56s200 89.7 200 200-89.7 200-200 200zm-80-216c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160-64c-17.7 0-32 14.3-32 32s14.3 32 32 32 32-14.3 32-32-14.3-32-32-32zm8 144H160c-13.2 0-24 10.8-24 24s10.8 24 24 24h176c13.2 0 24-10.8 24-24s-10.8-24-24-24z",
            ],
            "iconName": "meh",
            "prefix": "far",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-meh fa-w-16 "
          data-icon="meh"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 496 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 448c-110.3 0-200-89.7-200-200S137.7 56 248 56s200 89.7 200 200-89.7 200-200 200zm-80-216c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160-64c-17.7 0-32 14.3-32 32s14.3 32 32 32 32-14.3 32-32-14.3-32-32-32zm8 144H160c-13.2 0-24 10.8-24 24s10.8 24 24 24h176c13.2 0 24-10.8 24-24s-10.8-24-24-24z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </FontAwesomeIcon>
    </span>
  </div>
</RecommendationControl>
`;
